src/System.Memory/src/System/Buffers/ReadOnlyBuffer_helpers.cs

```diff
@@ -0,0 +1,305 @@
+                    {
+                        return new SequencePosition(start.Segment, startIndex + count);
+                    }
+                    return SeekMultiSegment((IMemoryList<byte>)start.Segment, startIndex, (IMemoryList<byte>)end.Segment, endIndex, count, checkEndReachable);
+
+                case BufferType.OwnedMemory:
+                case BufferType.Array:
+                    if (start.Segment != end.Segment)
+                    {
+                        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+                    }
+
+                    if (endIndex - startIndex >= count)
+                    {
+                        return new SequencePosition(start.Segment, startIndex + count);
+                    }
+
+                    ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();
+                    return default;
+
+                default:
+                    ThrowHelper.ThrowInvalidOperationException_UnexpectedSegmentType();
+                    return default;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal SequencePosition Seek(SequencePosition start, SequencePosition end, long count, bool checkEndReachable = true)
+        {
+            int startIndex = start.Index;
+            int endIndex = end.Index;
+            BufferType type = GetBufferType();
+
+            startIndex = GetIndex(startIndex);
+            endIndex = GetIndex(endIndex);
+
+            switch (type)
+            {
+                case BufferType.MemoryList:
+                    if (start.Segment == end.Segment && endIndex - startIndex >= count)
+                    {
+                        // end.Index >= count + Index and end.Index is int
+                        return new SequencePosition(start.Segment, startIndex + (int)count);
+                    }
+                    return SeekMultiSegment((IMemoryList<byte>)start.Segment, startIndex, (IMemoryList<byte>)end.Segment, endIndex, count, checkEndReachable);
+
+                case BufferType.OwnedMemory:
+                case BufferType.Array:
+                    if (endIndex - startIndex >= count)
+                    {
+                        // end.Index >= count + Index and end.Index is int
+                        return new SequencePosition(start.Segment, startIndex + (int)count);
+                    }
+
+                    ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();
+                    return default;
+
+                default:
+                    ThrowHelper.ThrowInvalidOperationException_UnexpectedSegmentType();
+                    return default;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static SequencePosition SeekMultiSegment(IMemoryList<byte> start, int startIndex, IMemoryList<byte> end, int endPosition, long count, bool checkEndReachable)
```

Comment:
This method can be re-written to reduce nesting, avoid unnecessary slicing, etc. You also don't need the foundResult flag.

@jkotas, @KrzysztofCwalina - thoughts?

Note: It is untested, so there might be some bugs that you may need to tweak.

```C#
private static SequencePosition SeekMultiSegment(IMemoryList<byte> start, int startIndex, IMemoryList<byte> end, int endPosition, long count, bool checkEndReachable)
{
    SequencePosition result = default;
    IMemoryList<byte> current = start;
    int currentIndex = startIndex;

    while (current != null)
    {
        bool isCurrentAtEnd = current == end;

        int memoryLength = isCurrentAtEnd ? endPosition - currentIndex : current.Memory.Length - currentIndex;

        if (memoryLength > count || (memoryLength == count && isCurrentAtEnd))
        {
            result = new SequencePosition(current, currentIndex + (int)count);
            break;
        }

        count -= memoryLength;
        currentIndex = 0;
        current = current.Next;
    }

    if (current == null)
        ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();

    if (checkEndReachable)
    {
        while (current != null)
            current = current.Next;
        if (current != end)
            ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
    }

    return result;
}
```